BUG: Preserve key order when using loc on MultiIndex DataFrame

nrebena opened the pull request:
## Description
closes #22797 
As described in #22797, the key order given to loc for a MultiIndex DataFrame was not respected:
```
import pandas as pd
import numpy as np
df = pd.DataFrame(np.arange(12).reshape((4, 3)),
    index=[['a', 'a', 'b', 'b'], [1, 2, 1, 2]],
    columns=[['Ohio', 'Ohio', 'Colorado'],
    ['Green', 'Red', 'Green']])

df.loc[(['b','a'],[2, 1]),:]

# Out
     Ohio     Colorado
    Green Red    Green
a 1     0   1        2
  2     3   4        5
b 1     6   7        8
  2     9  10       11
```

## Proposed fix
The culprit was the use of intersection of indexers in the loc function. I tried keeping the indexers sorted during the whole function (in the main loop), but performance were really affected (by a factor 3!!!).
As an other solution, I tried to sort the result after the indexers were computed. It was already way better (worse "only" by a factor 1.15 or so, see the asv benchmark result).
So I computed and add a flag testing if the result need to be sorted (the benchmark  seems to always have sorted key in the loc call).

**Update** The sorting function is now a separate private function (_reorder_indexer). It is called at the end of the get_locs function.

## Benchmark

Benchmark with the flag (I run asv compare with -s option):
<details>
Benchmarks that have got worse:

       before           after         ratio
     [39602e7d]       [da8b55af]
     <master>         <multiindex_sort_loc_order_issue_22797>
+      5.62±0.2μs       6.27±0.2μs     1.11  index_cached_properties.IndexCache.time_shape('Float64Index')
+      6.57±0.2μs       7.49±0.2μs     1.14  index_cached_properties.IndexCache.time_shape('TimedeltaIndex')
</details>

Benchmark without flag:
<details>
Benchmarks that have got worse:

       before           after         ratio
     [39602e7d]       [c786822a]
     <master>         <multiindex_sort_loc_order_issue_22797~1>
+     2.49±0.02ms      2.87±0.01ms     1.15  ctors.SeriesConstructors.time_series_constructor(<class 'list'>, False, 'int')
+        2.53±0ms      2.91±0.01ms     1.15  ctors.SeriesConstructors.time_series_constructor(<class 'list'>, True, 'int')
+      29.2±0.7ms      33.1±0.02ms     1.13  frame_ctor.FromLists.time_frame_from_lists
+        87.2±1ms         98.9±1ms     1.13  frame_ctor.FromRecords.time_frame_from_records_generator(None)
+     12.8±0.09ms      14.3±0.09ms     1.11  groupby.MultiColumn.time_col_select_numpy_sum
+      5.62±0.2μs       6.32±0.4μs     1.12  index_cached_properties.IndexCache.time_shape('Float64Index')
+     4.96±0.02ms      5.71±0.01ms     1.15  indexing.MultiIndexing.time_index_slice
+        2.91±0ms      3.29±0.01ms     1.13  inference.ToNumeric.time_from_numeric_str('coerce')
+        2.92±0ms      3.29±0.01ms     1.13  inference.ToNumeric.time_from_numeric_str('ignore')
+     3.45±0.01ms      3.84±0.01ms     1.11  series_methods.Map.time_map('lambda', 'object')
+      29.3±0.2ms      33.2±0.04ms     1.13  strings.Methods.time_len
</details>

## Checklist

- [x] closes #22797
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

nrebena committed:
TST: Test for issue #22797

Testing return order of MultiIndex.loc

MultiIndex.loc try to return the result in the same order as the key
given.

nrebena committed:
BUG: sort MultiIndex DataFrame loc result

From issue #22797. When given a list like object as indexer, the
returned result did not respect the order of the indexer, but the order
of the MultiIndex levels.

nrebena committed:
PERF: Skip sort of MultiIndex DataFrame loc result if not needed

Test if the result of the loc function need to be sorted to return them
in the same order as the indexer. If not, skip the sort to improve
performance.

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -3095,6 +3105,32 @@ def _update_indexer(idxr, indexer=indexer):
         if indexer is None:
             return Int64Index([])._ndarray_values
+
+        # Generate tuples of keys by which to order the results
```
this is really complex and adding quite a bit of code. Please take another look to simplify greatly.

nrebena replied:
Sure thing, will do. I did have a simpler solution, but the performance hit was really high.

nrebena committed:
CLN: Some code simplification

jreback reviewed (changes requested):
can we just *always* sort if the the level is already lexsorted, bascially putting your additional codes into that you added into _udpate_indexer?

nrebena commented:
This is kind of what I did in another PR #29190.
There is a bit of a performance drop on some benchmarking test.

To resume the two solution I made a PR for:
- In this PR: I do not modify the way the indexer is computed, and I sorted everything at the end.
- PR #29190: I keep the indexer sorted at all time, but with a bit of a performance drop for some benchmark, and improvement on other.

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -3095,6 +3103,31 @@ def _update_indexer(idxr, indexer=indexer):
             return Int64Index([])._ndarray_values
+
+        # Generate tuples of keys by which to order the results
+        if need_sort:
```
can you just check is_lexsorted?

nrebena replied:
This is not the same thing. The index may or may not be lexsorted, but what I want to know here is if the keys given to .loc are in the same order has the index (see line 3058), and if not, I reorder the result in indexer to have them in a order reflecting the given keys order.

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -3052,13 +3053,20 @@ def _update_indexer(idxr, indexer=indexer):
                 # are or'd)
+                # Find out if the list_like label are sorted as the levels or not
+                if not need_sort:
+                    k_codes = np.array(
```
use np.asarray and use .get_indexer (filter the -1 if you need to)

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -3095,6 +3103,31 @@ def _update_indexer(idxr, indexer=indexer):
         # empty indexer
         if indexer is None:
             return Int64Index([])._ndarray_values
+
```
can you move this to a separate function. Please type & document

nrebena committed:
CLN: Move code into separate function

Move code from get_locs to _reorder_indexer.
Better use of get_indexer to get level_code location.

TomAugspurger commented on pandas/core/indexes/multi.py:
```diff
@@ -3052,13 +3054,19 @@ def _update_indexer(idxr, indexer=indexer):
+                if not need_sort:
+                    k_codes = self.levels[i].get_indexer(k)
+                    k_codes = k_codes[k_codes >= 0]  # Filter absent keys
+                    need_sort = not (k_codes[:-1] < k_codes[1:]).all()
```
I don't recall: does NumPy short-circut `any`? If so it may be faster to change the comparision to `>=`, the `all` to an `any` and remove the `not` (haven't tested).

nrebena replied:
According to this discussion  https://stackoverflow.com/questions/45771554/why-numpy-any-has-no-short-circuit-mechanism#45773662, neither `all` nor `any` short-circuit anymore.
I will changed it however, as it approve readability.
Also, it point out an error I made, it should have been `<=` here, as two consecutive equals elements are still sorted. Nice catch.

nrebena committed:
CLN: More typing and linting

TomAugspurger commented on pandas/core/indexes/multi.py:
```diff
@@ -3095,8 +3103,54 @@ def _update_indexer(idxr, indexer=indexer):
+
+        Returns
+        -------
+        indexer : a sorted Int64Index of element of self ordered as seq
```
I don't understand "of element of self"

nrebena replied:
I mean an indexer of self. I will rephrase it.

TomAugspurger commented on pandas/core/indexes/multi.py:
```diff
@@ -3095,8 +3103,54 @@ def _update_indexer(idxr, indexer=indexer):
+
+        Parameters
+        ----------
+        seq : label/slice/list/mask or a sequence of such
```
We iterate over `seq`, so presumably it can't be a label or a slice. Can you confirm that it's just listlike or arraylike?

nrebena replied:
This is the same argument as in .get_locs, it is always a list, sometime with only one element, and the element can be a slice/label/list/mask for one level of the multi index.

jreback replied:
I think we just use ArrayLike here (can you try to type seq as well)

nrebena replied:
Ok for ArrayLike.
For seq, I logged the type passed when testing indexes and indexing, and it is a tuple that can contains the following type:
```
<class 'int'>
<class 'list'>
<class 'numpy.ndarray'>
<class 'pandas.core.indexes.numeric.Int64Index'>
<class 'pandas.core.series.Series'>
<class 'set'>
<class 'slice'>
<class 'str'>
<class 'tuple'>
```
So the type of seq would be something like 
`Tuple[Scalar, AnyArrayLike, Iterable, …] `
Does that make sense?

TomAugspurger commented on pandas/core/indexes/multi.py:
```diff
@@ -3095,8 +3103,54 @@ def _update_indexer(idxr, indexer=indexer):
+        for i, k in enumerate(seq):
+            if com.is_bool_indexer(k):
+                new_order = np.arange(n)[indexer]
+            elif is_list_like(k):
```
Is `k` an array? If so, pass `k.dtype` here.

nrebena replied:
k can be a slice, a mask, a single label or a list.
The only case to really handle here is a list, because it is the only kind of input that can be used in a different order than the level. For the other, it just ensure that it is not put out of order.

nrebena replied:
Most of the time (I mean in the example for multi index in the doc for example), k is a python list of label.

TomAugspurger reviewed (commented):
Can you add a whatsnew in 1.0.0.rst?

nrebena committed:
CLN: Improve readability and doc

nrebena committed:
DOC: Add a whatsnew

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -3095,8 +3105,56 @@ def _update_indexer(idxr, indexer=indexer):
+        -------
+        indexer : a sorted Int64Index indexer of self ordered as seq
+        """
+        from typing import Tuple
```
move to the top of module

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -3095,8 +3105,56 @@ def _update_indexer(idxr, indexer=indexer):
+        from typing import Tuple
+
+        n = len(self)
+        keys = tuple()  # type: Tuple[np.ndarray, ...]
```
you can use the py36 form here

nrebena replied:
Do you mean no type hint?
I add this when mypy complained with the following
`pandas/core/indexes/multi.py:3130: error: Need type annotation for 'keys'`

jreback commented on pandas/tests/indexes/multi/test_indexing.py:
```diff
@@ -438,3 +438,21 @@ def test_timestamp_multiindex_indexer():
+        columns=[["Ohio", "Ohio", "Colorado"], ["Green", "Red", "Green"]],
+    )
+
+    res = df.loc[["b", "a"], :]
```
can you also do 'a', 'b' on both of these cases

nrebena replied:
Done. I also have a additionnal question about a possible test case where I am not sure of the expected output.

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -3095,8 +3105,56 @@ def _update_indexer(idxr, indexer=indexer):
+            elif is_list_like(k):
+                # Generate a map with all level codes as sorted initially
+                key_order_map = np.ones(len(self.levels[i]), dtype=np.uint64) * len(
+                    self.levels[i]
```
I *think* that if you _sort_levels_monotonic then you don't need to reorder the indexers. is this possible?

nrebena replied:
The levels of the given index is maybe already monotonic, but we don't want to have the order as the index, but the same order as the keys in seq.
In the test case, the index are sorted, but we wanna return them in another order.
If I don't reorder the indexers, I will have to reorder the index itself before getting the indexer.
I did try this approach at the beginnig (reindexing then getting the indexer as previously done), but this is not possible due to the variety of type that can be given as key (slice, mask, array), because you can't reindex on a slice or a mask.

nrebena committed:
CLN: More typing for _reorder_indexer

nrebena commented:
When adding test cases, I come upon the following question.
What should be the ouput in this case
```
df = pd.DataFrame(
      np.arange(12).reshape((4, 3)),
      index=[["a", "a", "b", "b"], [1, 2, 1, 2]],
      columns=[["Ohio", "Ohio", "Colorado"], ["Green", "Red", "Green"]],
      )

df.loc[(slice(None), [1,2]), :]                                                                                                
# actual output
     Ohio     Colorado
    Green Red    Green
a 1     0   1        2
  2     3   4        5
b 1     6   7        8
  2     9  10       11

# maybe expected
     Ohio     Colorado
    Green Red    Green
a 2     3   4        5
b 2     9  10       11
a 1     0   1        2
b 1     6   7        8
```

nrebena committed:
TST: Add more test cases to test_multiindex_loc_order

jreback commented:
this got a bit lost, can you merge master and ping on green.

jreback commented on pandas/tests/indexes/multi/test_indexing.py:
```diff
@@ -438,3 +438,37 @@ def test_timestamp_multiindex_indexer():
     tm.assert_series_equal(result, should_be)
+
+
+def test_multiindex_loc_order():
```
can you move these to pandas/tests/test_multilevel.py

nrebena replied:
Yeah sure.

jreback commented on pandas/tests/indexes/multi/test_indexing.py:
```diff
@@ -438,3 +438,37 @@ def test_timestamp_multiindex_indexer():
+
+
+def test_multiindex_loc_order():
+    # GH 22797
```
if you think its easy to parameterize this would be great as well.

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -3095,8 +3106,56 @@ def _update_indexer(idxr, indexer=indexer):
+        indexer : a sorted Int64Index indexer of self ordered as seq
+        """
+        n = len(self)
+        keys = tuple()  # type: Tuple[np.ndarray, ...]
```
can you use explict typing (we dropped 3.5)

nrebena replied:
Done

jreback reviewed (changes requested):
can you merge master.

i think it might be worthile to remove need_sort entirely (from the main code) and just *always* call _reorder_indexer where you can determine the need_sort (and just bail early if you don't need it).

right?

then

nrebena committed:
TST: move test_multiindex_loc_order to tests/test_multilevel.py

nrebena committed:
CLN: Move need_sort test in _reorder_indexer

nrebena committed:
TST: Test also for columns

nrebena committed:
FIX: Test need sort only work on lexsorted indexes

nrebena commented:
> i think it might be worthile to remove need_sort entirely (from the main code) and just _always_ call _reorder_indexer where you can determine the need_sort (and just bail early if you don't need it).
This make the change more readable, thanks for the advice.

I merge master too.

I also added testing for columns selection.

WillAyd commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -908,6 +908,7 @@ MultiIndex
 
 - Constructior for :class:`MultiIndex` verifies that the given ``sortorder`` is compatible with the actual ``lexsort_depth``  if ``verify_integrity`` parameter is ``True`` (the default) (:issue:`28735`)
 - Series and MultiIndex `.drop` with `MultiIndex` raise exception if labels not in given in level (:issue:`8594`)
+- Bug in :meth:`Dataframe.loc` when used with a :class:`MultiIndex`. The returned values were not in the same order as the given inputs (:issue:`22797`)
```
Can you move this to 1.1 at this point?

nrebena replied:
Sure.

jreback commented on pandas/tests/test_multilevel.py:
```diff
@@ -2439,3 +2439,48 @@ def test_sort_ascending_list(self):
+        exp_index = pd.MultiIndex.from_arrays([["a", "a", "b", "b"], [1, 2, 1, 2]])
+        tm.assert_index_equal(res.index, exp_index)
+
+        res = df.loc[(["a", "b"], [1, 2]), :]
```
can you parameterize this, create additional tests for things that don't fit nicely

nrebena replied:
Done, but I fear i may have over parametrize…

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -3049,8 +3050,69 @@ def _update_indexer(idxr, indexer=indexer):
+                    # True if the given codes are not ordered
+                    need_sort = (k_codes[:-1] > k_codes[1:]).any()
+        # Bail out if no need to sort
+        # This is only true for a lexsorted index
```
can't we just test is_lexsorted?

jreback replied:
isn't that sufficient here?

nrebena replied:
The property we need to test is "do the index and the selection share the same order".
If we have 
```
df = pd.DataFrame(np.arange(2), index=[["b", "a"]]) 
df
   0
b  0
a  1

```
Then the index is not lexsorted, and we do not need to sort when doing
`df.loc[['b', 'a']]`, but we need to sort for getting `df.loc[['a', 'b']]` in the right order.

But they here room for improvement in this place, and I will propose something.

nrebena replied:
So I made a minor modification, but is the index is lexsorted, we still need to check if the requested order is also sorted before bailing out.

A another possibility is to sort in every case, if the performance hit is acceptable.

nrebena committed:
Minor change in how to determined if sorting is needed

nrebena committed:
PERF: Delete flag for sorting multiindex loc call.

nrebena committed:
TST: Parametrize tests

nrebena committed:
DOC: Move whatsnew entry from v1.0.0 to v1.1.0

nrebena committed:
Revert "PERF: Delete flag for sorting multiindex loc call."

This reverts commit 7fee53c6234e20830f61ab096b332f8c7e421359.

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -114,7 +114,7 @@ Missing
 MultiIndex
 ^^^^^^^^^^
-
+- Bug in :meth:`Dataframe.loc` when used with a :class:`MultiIndex`. The returned values were not in the same order as the given inputs (:issue:`22797`)
```
can you make a mini-example (separate sub-section). I think this is very hard to visualize from the text, but the basic example we are using is very clear.

nrebena replied:
Sure thing.

jreback reviewed (changes requested):
looks good. if you can update the whatsnew with an example and rebase can merge this in. sorry take a while, this is tricky code you are fixing.

nrebena committed:
DOC: Add mini example to whatsnew

jreback commented:
@nrebena sorry, if you'd merge master again. ping on green.

nrebena commented:
@jreback Merged and green

jreback commented:
thanks @nrebena very nice! sorry took a long time :>

nrebena commented:
@jreback It was a pleasure, I learned a lot. Thank for the guidance.
